src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,256 @@
+///! Driver relays driver
+///!
+///! There are 2 relays at the driver output:
+///!    - one that shorts the output to ground
+///!    - one that connects the current source/sink to the output
+///!
+///! See [Relay] documentation for details about relay control.
+///!
+///! The relays are controlled via an MCP23008 I2C io expander.
+use super::hal::rcc;
+use core::fmt::Debug;
+use embedded_hal::blocking::{
+    delay::DelayUs,
+    i2c::{Write, WriteRead},
+};
+use mcp230xx::{Level, Mcp23008, Mcp230xx};
+
+use super::Channel;
+use smlang::statemachine;
+
+// Driver output relays pins.
+#[allow(non_camel_case_types)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum RelayPin {
+    LN_K1_EN_N,
+    LN_K1_EN,
+    LN_K0_D,
+    LN_K0_CP,
+    HP_K1_EN_N,
+    HP_K1_EN,
+    HP_K0_D,
+    HP_K0_CP,
+}
+
+impl From<RelayPin> for Mcp23008 {
+    fn from(pin: RelayPin) -> Self {
+        match pin {
+            RelayPin::LN_K1_EN_N => Self::P0,
+            RelayPin::LN_K1_EN => Self::P1,
+            RelayPin::LN_K0_D => Self::P2,
+            RelayPin::LN_K0_CP => Self::P3,
+            RelayPin::HP_K1_EN_N => Self::P4,
+            RelayPin::HP_K1_EN => Self::P5,
+            RelayPin::HP_K0_D => Self::P6,
+            RelayPin::HP_K0_CP => Self::P7,
+        }
+    }
+}
+
+// small helper to lock the mutex
+// maybe todo: Pass out an error if in use. Not sure how to get that out of the state machine though..
+fn get_mcp<I2C>(
+    mutex: &'_ spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+) -> spin::MutexGuard<Mcp230xx<I2C, Mcp23008>> {
+    mutex.try_lock().unwrap() // panic here if in use
+}
+
+/// Driver output [Relay] type that controlls the set of two relays that is present on both of
+/// the driver output channels.
+/// [Relay] contains a mutex of a shared MCP23008, an I2C io expander chip with 8 GPIOs.
+/// 4 of those GPIOs are used to controll one set of relays (aka one [Relay]):
+///     - `k1_en_n` and `k1_en` differentially control the state of K1
+///     - `k0_d` is that data input of a flipflop which controls the state of K0
+///     - `k0_cp` is the clock input if the flipflop. Note that there is additional circuitry
+///        which ensures this clock input cannot be driven if K1 is not in the correct state.
+///        See Driver schematic for exact details.
+/// A `delay` is used to generate a rising edge for the flipflip with enaugh timing margins.
+pub struct Relay<'a, I2C: WriteRead + Write> {
+    mutex: &'a spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
```

Comment:
You should (but I've been wrong before) be able to get rid of the lifetime parameter. Since all you support is static anyways.

```suggestion
pub struct Relay<I2C: WriteRead + Write> {
    mutex: &'static spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
```

And give it a better name than `mutex`. E.g. `gpio`, because the inner value is the relevant one, not the mutex.